core/src/main/java/org/elasticsearch/action/bulk/Retry.java

```diff
@@ -0,0 +1,226 @@
+                addResponses(bulkItemResponses, (r -> true));
+                BulkResponse aggregatedResults = getCompleteResponse();
+                notifyListener(aggregatedResults);
+            } else {
+                if (canRetry(bulkItemResponses)) {
+                    addResponses(bulkItemResponses, (r -> !r.isFailed()));
+                    retry(createBulkRequestForRetry(bulkItemResponses));
+                } else {
+                    addResponses(bulkItemResponses, (r -> true));
+                    BulkResponse aggregatedResults = getCompleteResponse();
+                    notifyListener(aggregatedResults);
+                }
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            try {
+                listener.onFailure(e);
+            } finally {
+                s.shutdownNow();
+            }
+        }
+
+        private void retry(BulkRequest bulkRequestForRetry) {
+            if (backoff.hasNext()) {
+                s.schedule((() -> this.execute(bulkRequestForRetry)), backoff.next().millis(), TimeUnit.MILLISECONDS);
+            }
+        }
+
+        private BulkRequest createBulkRequestForRetry(BulkResponse bulkItemResponses) {
+            BulkRequest requestToReissue = new BulkRequest();
+            int index = 0;
+            for (BulkItemResponse bulkItemResponse : bulkItemResponses.getItems()) {
+                if (bulkItemResponse.isFailed()) {
+                    // Assumption: ES maintains order between request and response
+                    requestToReissue.add(currentBulkRequest.requests().get(index));
+                }
+                index++;
+            }
+            return requestToReissue;
+        }
+
+        private boolean canRetry(BulkResponse bulkItemResponses) {
+            if (!backoff.hasNext()) {
+                return false;
+            }
+            for (BulkItemResponse bulkItemResponse : bulkItemResponses) {
+                if (bulkItemResponse.isFailed()) {
+                    Throwable cause = bulkItemResponse.getFailure().getCause();
+                    Throwable rootCause = ExceptionsHelper.unwrapCause(cause);
+                    if (!rootCause.getClass().equals(retryOnThrowable)) {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+
+        private void notifyListener(BulkResponse response) {
+            try {
+                listener.onResponse(response);
+            } finally {
+                s.shutdownNow();
+            }
+        }
+
+        private void addResponses(BulkResponse response, Predicate<BulkItemResponse> filter) {
+            for (BulkItemResponse bulkItemResponse : response) {
+                if (filter.test(bulkItemResponse)) {
+                    responses.add(bulkItemResponse);
+                }
+            }
+        }
+
+        private BulkResponse getCompleteResponse() {
```

Comment:
Fine by me. I tend not to use `get` unless I'm implementing a method to get a property just because getters are so common in java. If I ever think of the method as building a thing out of state  then I don't use `get`. Its kind of silly because the whole point of getters is that you can build the thing. `getAccumulatedResponse` is fine.